AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/SubscriptionSync/ReconcileAndLocalSave/RemoteSyncReconciler.swift

```diff
@@ -12,39 +12,102 @@ import Amplify
-    typealias SavedModel = ReconcileAndLocalSaveOperation.AppliedModel
 
     enum Disposition {
-        case applyRemoteModel(RemoteModel, MutationEvent.MutationType)
-        case dropRemoteModel(String)
+        case create(RemoteModel)
+        case update(RemoteModel)
+        case delete(RemoteModel)
     }
 
-    static func reconcile(remoteModel: RemoteModel,
-                          to localMetadata: LocalMetadata?,
-                          pendingMutations: [MutationEvent]) -> Disposition {
+    /// Reconciles the incoming `remoteModels` against the pending mutations.
+    /// If there is a matching pending mutation, drop the remote model.
+    ///
+    /// - Parameters:
+    ///   - remoteModels: models retrieved from the remote store
+    ///   - pendingMutations: pending mutations from the outbox
+    /// - Returns: remote models to be applied
+    static func reconcile(_ remoteModels: [RemoteModel],
+                          pendingMutations: [MutationEvent]) -> [RemoteModel] {
+        guard !pendingMutations.isEmpty else {
+            return remoteModels
+        }
+
+        let pendingMutationModelIdsArr = pendingMutations.map { mutationEvent in
+            mutationEvent.modelId
+        }
+        let pendingMutationModelIds = Set(pendingMutationModelIdsArr)
+
+        return remoteModels.filter { remoteModel in
+            !pendingMutationModelIds.contains(remoteModel.model.id)
+        }
+    }
 
-        guard pendingMutations.isEmpty else {
-            return .dropRemoteModel(remoteModel.model.modelName)
+    /// Reconciles the incoming `remoteModels` against the local metadata.
+
+    ///
+    /// - Parameters:
+    ///   - remoteModels: models retrieved from the remote store
+    ///   - localMetadatas: metadata retrieved from the local store
+    /// - Returns: disposition of models to apply locally
+    static func reconcile(_ remoteModels: [RemoteModel],
+                          localMetadatas: [LocalMetadata]) -> [Disposition] {
+        var dispositions = [Disposition]()
+
+        guard !remoteModels.isEmpty else {
+            return dispositions
+        }
+
+        guard !localMetadatas.isEmpty else {
+            remoteModels.forEach { remoteModel in
+                if let disposition = reconcile(remoteModel, localMetadata: nil) {
+                    dispositions.append(disposition)
+                }
+            }
+            return dispositions
         }
 
+        var localMetadataMap = [Model.Identifier: LocalMetadata]()
+        for localMetadata in localMetadatas {
+            localMetadataMap.updateValue(localMetadata, forKey: localMetadata.id)
+        }
+
+        remoteModels.forEach { remoteModel in
+            if let disposition = reconcile(remoteModel, localMetadata: localMetadataMap[remoteModel.model.id]) {
+                dispositions.append(disposition)
+            }
+        }
+
+        return dispositions
```

Comment:
Can simplify this with std lib array methods:
```suggestion
        guard !remoteModels.isEmpty else {
            return []
        }

        guard !localMetadatas.isEmpty else {
            let dispositions = remoteModels.compactMap { reconcile($0, localMetadata: nil) }
            return dispositions
        }

        let metadataByID = localMetadatas.reduce(into: [:]) { $0[$1.id] = $1 }

        let dispositions = remoteModels.compactMap { reconcile($0, localMetadata: metadataByID[$0.model.id]) }

        return dispositions
```